cdap-kms/src/main/java/co/cask/cdap/security/store/KMSSecureStore.java

```diff
@@ -0,0 +1,148 @@
+   * @param data The data that needs to be securely stored.
+   * @param description User provided description of the entry.
+   * @param properties Metadata associated with the data
+   * @throws IOException If it failed to store the key in the store.
+   */
+  @Override
+  public void put(String namespace, String name, byte[] data, String description, Map<String, String> properties)
+    throws IOException {
+    KeyProvider.Options options = new KeyProvider.Options(conf);
+    options.setDescription(description);
+    options.setAttributes(properties);
+    options.setBitLength(data.length * Byte.SIZE);
+    try {
+      provider.createKey(SecureStoreUtils.getKeyName(namespace, name), data, options);
+    } catch (IOException e) {
+      throw new IOException("Failed to store the key. ", e);
+    }
+  }
+
+  /**
+   * Deletes the element with the given name.
+   * @param namespace The namespace this key belongs to.
+   * @param name Name of the element to be deleted.
+   */
+  @Override
+  public void delete(String namespace, String name) throws IOException {
+    try {
+      provider.deleteKey(SecureStoreUtils.getKeyName(namespace, name));
+    } catch (IOException e) {
+      throw new IOException("Failed to delete the key. ", e);
+    }
+  }
+
+  /**
+   * List of all the entries in the secure store.
+   * @return A list of {@link SecureStoreMetadata} objects representing the data stored in the store.
+   * @param namespace The namespace this key belongs to.
+   */
+  @Override
+  public List<SecureStoreMetadata> list(String namespace) throws IOException {
+    String prefix = namespace + SecureStoreUtils.NAME_SEPARATOR;
+    List<String> keysInNamespace = new ArrayList<>();
+    KeyProvider.Metadata[] metadatas;
+    List<SecureStoreMetadata> secureStoreMetadatas;
+    try {
+      for (String key : provider.getKeys()) {
+        if (key.startsWith(prefix)) {
+          keysInNamespace.add(key);
+        }
+      }
+      metadatas = provider.getKeysMetadata(keysInNamespace.toArray(new String[0]));
+      secureStoreMetadatas = new ArrayList<>();
+    } catch (IOException e) {
+      throw new IOException("Failed to get the list of elements from the secure store.", e);
+    }
+    for (int i = 0; i < metadatas.length; i++) {
+      KeyProvider.Metadata metadata = metadatas[i];
+      SecureStoreMetadata meta = SecureStoreMetadata.of(keysInNamespace.get(i).substring(prefix.length()),
+                                                        metadata.getDescription(),
+                                                        metadata.getAttributes());
+      secureStoreMetadatas.add(meta);
+    }
+    return secureStoreMetadatas;
+  }
+
+  /**
+   * @param namespace The namespace this key belongs to.
+   * @param name Name of the data element.
+   * @return An object representing the securely stored data associated with the name.
+   */
+  @Override
+  public SecureStoreData get(String namespace, String name) throws IOException {
```

Comment:
Unfortunately this is what we are stuck with give the KMS APIs. I will open a JIRA and later try to look for a better way to present these errors to the user.

> On Jul 17, 2016, at 10:00 AM, Bhooshan Mogal <notifications@github.com> wrote:
> 
> In cdap-kms/src/main/java/co/cask/cdap/security/store/KMSSecureStore.java <https://github.com/caskdata/cdap/pull/6066#discussion_r71085591>:
> 
> > +      KeyProvider.Metadata metadata = metadatas[i];
> > +      SecureStoreMetadata meta = SecureStoreMetadata.of(keysInNamespace.get(i).substring(prefix.length()),
> > +                                                        metadata.getDescription(),
> > +                                                        metadata.getAttributes());
> > +      secureStoreMetadatas.add(meta);
> > +    }
> > +    return secureStoreMetadatas;
> > +  }
> > +
> > +  /**
> > +   * @param namespace The namespace this key belongs to.
> > +   * @param name Name of the data element.
> > +   * @return An object representing the securely stored data associated with the name.
> > +   */
> > +  @Override
> > +  public SecureStoreData get(String namespace, String name) throws IOException {
> But the REST API will then respond with a 500, and clients (specifically UI) will have to rely on the error message.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/caskdata/cdap/pull/6066/files/e87907bc48cccf7b7eb520b1f5e6ff174ed59113#r71085591>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAFGk2FK7UzDUkfqlTMevekMEhDBQY3Oks5qWl-RgaJpZM4JOEs0>.
>